sigstore/_internal/rekor/client.py

```diff
@@ -219,14 +193,86 @@ def post(
 
-        resp: requests.Response = self.session.post(self.url, data=json.dumps(data))
+        resp: requests.Response = self.session.post(self.url, json=data)
         try:
             resp.raise_for_status()
         except requests.HTTPError as http_error:
             raise RekorClientError from http_error
 
         return RekorEntry.from_response(resp.json())
 
+    @property
+    def retrieve(self) -> RekorEntriesRetrieve:
+        return RekorEntriesRetrieve(
+            urljoin(self.url, "retrieve/"), session=self.session
+        )
+
+
+class RekorEntriesRetrieve(Endpoint):
+    def post(
+        self,
+        b64_artifact_signature: str,
+        sha256_artifact_hash: str,
+        b64_cert: str,
+    ) -> Optional[RekorEntry]:
+        """
+        Retrieves an extant Rekor entry, identified by its artifact signature,
+        artifact hash, and signing certificate.
+
+        Returns None if Rekor has no entry corresponding to the signing
+        materials.
+        """
+        data = {
+            "entries": [
+                {
+                    "kind": "hashedrekord",
+                    "apiVersion": "0.0.1",
+                    "spec": {
+                        "signature": {
+                            "content": b64_artifact_signature,
+                            "publicKey": {"content": b64_cert},
+                        },
+                        "data": {
+                            "hash": {
+                                "algorithm": "sha256",
+                                "value": sha256_artifact_hash,
+                            }
+                        },
+                    },
+                }
+            ]
+        }
+
+        resp: requests.Response = self.session.post(self.url, json=data)
+        try:
+            resp.raise_for_status()
+        except requests.HTTPError as http_error:
+            if http_error.status_code == 404:
+                return None
+            raise RekorClientError(resp.json()) from http_error
+
+        # Rekor should return a 404 when there are no results rather than
+        # an empty list, but we check for the latter just in case.
+        #
+        # Similarly, we expect exactly one top-level response, since we made
+        # exactly one entry request. The inner entry set may have more than
+        # one result, but anything other than 1 at the top-level indicates
+        # a bug in Rekor.
+        body = resp.json()
+        if len(body) == 0:
+            return None
+        elif len(body) > 1:
+            raise RekorClientError(
+                f"expected exactly one response for entry query, but got {len(body)}"
+            )
+
+        # The response is a list of `{ uuid: LogEntry }` objects.
+        # We select the oldest entry to actually return, since a malicious
+        # actor could conceivably spam the log with newer duplicate entries.
+        entry = min(body[0].items(), key=lambda tup: tup[1]["integratedTime"])
```

Comment:
This is a little ugly, but it accomplishes the filter without surfacing it beyond the client API. Thoughts @asraa @di?